dali/operators/audio/nonsilence_op.h

```diff
@@ -0,0 +1,250 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_AUDIO_NONSILENCE_OP_H_
+#define DALI_OPERATORS_AUDIO_NONSILENCE_OP_H_
+
+#include <cmath>
+#include <limits>
+#include <memory>
+#include <utility>
+#include <vector>
+#include "dali/core/convert.h"
+#include "dali/kernels/kernel_manager.h"
+#include "dali/kernels/signal/decibel/decibel_calculator.h"
+#include "dali/kernels/signal/moving_mean_square.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/pipeline/operator/operator.h"
+
+namespace dali {
+namespace detail {
+
+const int kNumOutputs = 2;
+using OutputType = int;
+static_assert(std::is_integral<OutputType>::value,
+              "Operator return indices, thus OutputType shall be integral");
+
+}  // namespace detail
+
+template<typename Backend>
+class NonsilenceOperator : public Operator<Backend> {
+ public:
+  ~NonsilenceOperator() override = default;
+
+  DISABLE_COPY_MOVE_ASSIGN(NonsilenceOperator);
+
+ protected:
+  explicit NonsilenceOperator(const OpSpec &spec) :
+          Operator<Backend>(spec) {}
+
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+
+  void AcquireArgs(const OpSpec &spec, const workspace_t<Backend> &ws) {
+    this->GetPerSampleArgument(cutoff_db_, "cutoff_db", ws);
+    this->GetPerSampleArgument(reference_db_, "reference_db", ws);
+    this->GetPerSampleArgument(reference_max_, "reference_max", ws);
+    this->GetPerSampleArgument(window_length_, "window_length", ws);
+    this->GetPerSampleArgument(reset_interval_, "reset_interval", ws);
+    auto input_type = ws.template InputRef<Backend>(0).type().id();
+    if (!IsFloatingPoint(input_type)) {
+      // If input type is integral, no need for reset interval
+      reset_interval_.assign(reset_interval_.size(), -1);
+    }
+  }
+
+
+  std::vector<float> cutoff_db_;
+  std::vector<float> reference_db_;
+  std::vector<bool> reference_max_;
+  std::vector<int> window_length_;
+  std::vector<int> reset_interval_;
+
+  USE_OPERATOR_MEMBERS();
+};
+
+
+class NonsilenceOperatorCpu : public NonsilenceOperator<CPUBackend> {
+ public:
+  explicit NonsilenceOperatorCpu(const OpSpec &spec) :
+          NonsilenceOperator<CPUBackend>(spec),
+          impl_(std::make_unique<Impl>(this)) {}
+
+
+  ~NonsilenceOperatorCpu() override = default;
+
+  DISABLE_COPY_MOVE_ASSIGN(NonsilenceOperatorCpu);
+
+  class Impl;
+
+ protected:
+  bool SetupImpl(std::vector<::dali::OutputDesc> &output_desc,
+                 const workspace_t<CPUBackend> &ws) override;
+
+  void RunImpl(workspace_t<CPUBackend> &ws) override;
+
+ private:
+  std::unique_ptr<Impl> impl_;
+};
+
+
+class DLL_PUBLIC NonsilenceOperatorCpu::Impl {
+ private:
+  using MmsArgs = kernels::signal::MovingMeanSquareArgs;
+
+ public:
+  Impl() = default;
+
+
+  explicit Impl(const NonsilenceOperatorCpu *enclosing) : enclosing_(enclosing),
```

Comment:
I think it would be better to just get the fields you need and don't store the pointer to the operator instance